Only register singletons in the service locator

blairconrad opened the pull request:
… and resolve them eagerly.

Connects to #1478.

blairconrad commented on tests/FakeItEasy.Tests/FakeTests.cs:
```diff
@@ -1,9 +1,71 @@
+                Expression<Action> call = () => Fake.InitializeFixture(new object());
+#pragma warning restore CS0618 // Type or member is obsolete
+                call.Should().BeNullGuarded();
+        }
```
All the new tests here were sucked over from the `FakeFacadeTests`. Some went to the specs.

blairconrad commented:
There are other potential improvements as well. If nothing else, we register a lot of services only so we can find them inside the *Modules, and they have no external consumers. While I was chasing that path, I had difficulties with some of the differences between the singletons and the transient "services", and thought this change would be relatively self-contained.

thomaslevesque commented on src/FakeItEasy/FakeFacade.cs:
```diff
@@ -1,65 +0,0 @@
-#pragma warning restore CS0618 // Type or member is obsolete
-        }
-    }
-}
```
Good riddance! I never understood the need for this class

thomaslevesque commented on tests/FakeItEasy.Tests/Core/DefaultFakeManagerAccessorTests.cs:
```diff
@@ -10,12 +10,7 @@
-        {
-            this.accessor = new DefaultFakeManagerAccessor();
-        }
+        private readonly IFakeManagerAccessor accessor = ServiceLocator.Current.Resolve<IFakeManagerAccessor>();
```
Are you sure about this change? This test fixture is `DefaultFakeManagerAccessorTests`, but it never explicitly creates a `DefaultFakeManagerAccessor`; instead it relies on the service locator to return a `DefaultFakeManagerAccessor`. This seems wrong...

blairconrad replied:
I am not sure at all. I have no idea why I did that. (Nor any memory of it, but that's not unusual.) Thank you!

thomaslevesque commented on src/FakeItEasy/RootModule.cs:
```diff
@@ -106,7 +106,7 @@ public static void RegisterDependencies(DictionaryContainer container)
             container.RegisterSingleton(c => new EventHandlerArgumentProviderMap());
 
-            container.Register(c => new SequentialCallContext(c.Resolve<CallWriter>(), c.Resolve<StringBuilderOutputWriter.Factory>()));
+            container.RegisterSingleton<SequentialCallContext.Factory>(c => () => new SequentialCallContext(c.Resolve<CallWriter>(), c.Resolve<StringBuilderOutputWriter.Factory>()));
```
Small optimization: resolve the `CallWriter` and `StringBuilderOutputWriter.Factory` outside of the inner delegate, i.e.

```csharp
container.RegisterSingleton<SequentialCallContext.Factory>(c =>
{
    var callWriter = c.Resolve<CallWriter>();
    var outputWriterFactory = c.Resolve<StringBuilderOutputWriter.Factory>();
    return () => new SequentialCallContext(callWriter, outputWriterFactory);
});
```

This could probably be done in other places, too. Not sure how big the performance impact would be, but I guess it can't hurt.

thomaslevesque replied:
In fact, ignore this comment. I had forgotten that it was a singleton, so it will only be constructed once anyway.

blairconrad replied:
Also, those changes will come in time!

thomaslevesque commented on src/FakeItEasy/IoC/DictionaryContainer.cs:
```diff
@@ -8,120 +8,69 @@ namespace FakeItEasy.IoC
-                    this.resolveFunction = null;
-                }
-            }
+            throw new KeyNotFoundException($"The specified service {serviceType} was not registered in the container.");
```
Shouldn't there be some locking in this method? If two threads attempt to modified `cachedServices` at the same time, it could become corrupted. Or maybe use a `ConcurrentDictionary`?

thomaslevesque replied:
And... ignore this too. All services are constructed eagerly from a single thread, and the dictionary won't be modified afterward.

blairconrad replied:
It's currently safe, since we're only accessing the `DictionaryContainer` from the static constructor of `ServiceLocator`, which is guaranteed to run once. If you want futureproofing, I can add.

thomaslevesque commented on src/FakeItEasy/ServiceLocator.cs:
```diff
@@ -1,30 +1,45 @@
-        [DebuggerStepThrough]
-        internal abstract object Resolve(Type componentType);
+            throw new KeyNotFoundException($"The specified service {typeof(T)} was not registered.");
+        }
```
Nice! I like how you removed the inheritance between `DictionaryContainer` and `ServiceLocator`.

blairconrad commented:
Thanks for the insightful comments, @thomaslevesque. I think I addressed (reverted) the one change you were unhappy with. Also, I rebased and took the liberty of pushing a bonus commit from the future. It stops us using `ServiceLocator` from within `RootModule` code, and is a preview of how we can stop registering many types used only inside the *Modules.

blairconrad commented:
Sorry, I just amended 72b9a1bf574b55f6afdbd176b41830f92ef35379 instead of including the fixup commit (dumb/lazy), but the only change I made was to revert the area you called out, and am now (once again) creating a `DefaultFakeManagerAccessor` in its test class.

blairconrad commented:
Thank you, @thomaslevesque! More changes to come in another PR (not today). Unfortunately, the next phase is going to be a bit of a slog, while I switch from registering and resolving everything to eagerly instantiating objects and then only registering if we need to.
